### PR TITLE
llvm-core: bump version to 12.0.0

### PR DESCRIPTION
**Pull request by newlawrence**
Specify library name and version:  **llvm-core/12.0.0**

This is the newest version of the **LLVM** core libraries released on April the 15th.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
This might get you https://github.com/conan-io/conan-center-index/blob/a3d762103210991baa4037a563938fe069016467/recipes/llvm-core/all/conanfile.py#L69

**Comment by newlawrence** [comment]
Terminfo was disabled in the latest commit as it was causing some issues depending on the host toolchain. As it is only needed by the executables and they are not deployed in the package (only libraries are kept), it can be disabled without any impact.

**Comment by newlawrence** [comment]
> This might get you
> 
> https://github.com/conan-io/conan-center-index/blob/a3d762103210991baa4037a563938fe069016467/recipes/llvm-core/all/conanfile.py#L69

You're right, additional folders for the patches have been added... but it has worked for me locally. Let's wait for the **CI** and see.

**Comment by prince-chrismc** [comment]
> Let's wait for the **CI** and see.

It passed... Conan magic?

**Comment by newlawrence** [comment]
> It passed... Conan magic?

It seems so 😁.

Now let's see if I haven't break anything by fixing **MSBuild** builds. Usually I test the recipes in **Windows** using **Ninja** (which is way better). The luck I had with `export_sources` I didn't had it while manually copying one of the libraries that refuses to get installed by standard **CMake** means.

**Comment by prince-chrismc** [comment]
You'll need to retrigger CI, close the pr wait 10s and then re-open it 🔁

**Comment by prince-chrismc** [comment]
You might want to block MT since it does not seem to be supported

**Comment by newlawrence** [comment]
> You might want to block MT since it does not seem to be supported

Is it stated somewhere in the **LLVM** docs? By looking at the **CI** results I don't think that is the case (if you take a look [here](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/5720/3/llvm-core/12.0.0//summary.json) you'll see that both, **MT** and **MD** failed). However, good point that you've mentioned the **CRT** flags. I missed the `LLVM_USE_CRT_*` flag from the **LLVM** documentation; it is now [fixed](https://github.com/conan-io/conan-center-index/pull/5720/commits/e44d309198c1606369ef6fb7617689ad564f1f09) 😊.

I also seized the oportunity to get rid of a patch that seems to be no longer necessary since version **11**.

However, I don't think these two changes are going to make the **CI** pass. If you take a look at the **CMake** logs of one of the tests on **Windows** ([this one](https://c3i.jfrog.io/c3i/misc/logs/pr/5720/4/llvm-core/12.0.0//ae557f160d180423986df18d1ca7d364dc0024a7-test.txt), for instance) and compare it to another one on **Linux** (let's say, [this one](https://c3i.jfrog.io/c3i/misc/logs/pr/5720/4/llvm-core/12.0.0//d1405139a41ea2a6c5bdf7038aa86f5382092ab2-test.txt)), you'll notice that the vast majority of the libraries are not mentioned as founded by **CMake** on **Windows** but, if you take a look at the correspoding build logs, on both platforms the same 154 static libraries are built. That means that what are wrong are the **Conan** components in the `package_info` method; we're missing a lot of them.

I don't know yet what can be causing the issue. So far I haven't been able to reproduce it locally. What is curious is that it is something that is only happening on version **12**. I'll keep investigating...

**Comment by newlawrence** [comment]
There's definitely an issue with the **graphviz** dot file I use to model the package components as can be seen by comparing the logs for 11 and 12 (I've disabled all configurations but **Windows** ones to avoid extra noise).

The graph for version 12 showed below explains why there are missing symbols while building the tests. It is missing pretty much everything from the library.

```json
{
    "LLVMSupport": ["advapi32", "delayimp", "ole32", "psapi", "shell32", "uuid"],
    "LLVMFileCheck": ["LLVMSupport"],
    "LLVMFrontendOpenACC": ["LLVMSupport"],
    "LLVMObject": [],
    "LLVMJITLink": ["LLVMObject",
    "LLVMOrcTargetProcess",
    "LLVMSupport"],
    "LLVMOrcTargetProcess": [],
    "LLVMAnalysis": [],
    "LLVMOrcJIT": ["LLVMAnalysis",
    "LLVMBitReader",
    "LLVMBitWriter",
    "LLVMPasses"],
    "LLVMBitReader": [],
    "LLVMBitWriter": [],
    "LLVMPasses": [],
    "LLVMTableGenGlobalISel": ["LLVMSupport", "LLVMTableGen"],
    "LLVMTableGen": []
}
```

Why is this happening? I've no idea. Parsing the resulting **graphviz** file works for the other configurations, including the previous **Windows** version, which I think it discards the idea of it being an issue related to the **CMake** version. As I told before, I cannot even reproduce it locally.

These are the lines in the `conanfile.py` that I use to generate and load the dot file in case they're of any help. Maybe there is **CMake** flag that can help fixing the issue?

```python
self.run('cmake --graphviz=graph/llvm.dot .')
with tools.chdir('graph'):
    dot_text = tools.load('llvm.dot').replace('\r\n', '\n')
```

**Comment by newlawrence** [comment]
I'm summoning @jgsogo here.

**Comment by jgsogo** [comment]
It is the CMake version. I don't know what has changed on the LLVM project (I was pointed to https://github.com/llvm/llvm-project/commit/9218ff50f93085d0a16a974db28ca8f14bc66f64) or on the CMake side, but using CMake 3.20.2 it generates the expected graph:

https://github.com/conan-io/conan-center-index/pull/5885#issuecomment-860754802

You will need to add `cmake` as build-requires, as we prefer to keep the CMake version in the CI more aligned with [the one required by Conan itself](https://github.com/conan-io/tribe/blob/main/design/004-tools-cmake.md).

**Comment by newlawrence** [comment]
Done @jgsogo. Thanks for the help!

File: recipes/llvm-core/all/conanfile.py [hunk 1]
```diff
@@ -65,6 +65,9 @@ class LLVMCoreConan(ConanFile):
         'with_zlib': True,
         'with_xml2': True
     }
+    build_requires = [
+        'cmake/3.20.3'
+    ]
```
Comment (jgsogo) [inline recipes/llvm-core/all/conanfile.py:70]: Please, add a comment close to this build-requirement to tell why it is needed. Our future us will be grateful to our today-us 😄

**Review by jgsogo — CHANGES_REQUESTED** [review]
IMHO, adding that comment to the recipe is really important.

**Comment by jgsogo** [comment]
Also, please, as we are already starting to build [binaries for M1](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/5720/12-tapaholes/llvm-core/12.0.0//summary.json) (not public in PRs yet), better if we start to consider them. I feel like this recipe is not going to be easy to cross-compile, so just add the following to it to skip that configuration:

```python
    def validate(self):
        if tools.cross_building(self, skip_x64_x86=True):
            raise ConanInvalidConfiguration("Cross-building not implemented")
```

File: recipes/llvm-core/all/conanfile.py (hunk 1 above)
Comment (newlawrence) [reply recipes/llvm-core/all/conanfile.py:73]: [Done](https://github.com/conan-io/conan-center-index/pull/5720/commits/3ac945755c6900b1584dd82d2c5da4df45cc4aae)!

**Comment by newlawrence** [comment]
> Also, please, as we are already starting to build [binaries for M1](https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/5720/12-tapaholes/llvm-core/12.0.0//summary.json) (not public in PRs yet), better if we start to consider them. I feel like this recipe is not going to be easy to cross-compile.

Well, the main difficulty in cross compiling **LLVM** is that running the generated **llvm-tblgen** executable is a step in the process. You can use whatever version of it you have in the host machine by setting `LLVM_TABLEGEN` though.

If this package gets successful I may create a package for **llvm-tblgen** just to use it as a build requirement for cross compiling. For the moment I've raised `ConanInvalidConfiguration` as you suggested @jgsogo.

**Review by jgsogo — APPROVED** [review]
One step closer 😃

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**